[RF DOCS] Action Mailbox Documention [ci-skip]

bhumi1102 opened the pull request:
## Overview

Action Mailbox is a pretty useful / interesting sub-framework of Rails, that more apps probably could benefit from. It does have a number of moving parts so I focused on explaining the flow of how things work. I did a lot of 'in-app testing' of the docs by implementing its features in a rails app and exploring what's missing in the docs.

## New sections

Here are the new things I added, in addition to clarifying the existing text.

- Processing Incoming Email - more of a step by step of the flow
- Mail object and it's fields
- List of Inbound Email 'status' values
- ActiveJobs that are part of Action Mailbox - Routing and Incineration

## Testing

In addition to testing the Action Mailbox features in-app, I also ran `guides:generate` and `guides:validate` locally. I perused the generated html locally as well.

I also tested the instructions for one 'real' ingress Postmark (locally with ngrok).

bhumi1102 committed:
section 1

bhumi1102 committed:
section 2

bhumi1102 committed:
section 3 processing email

bhumi1102 committed:
incineration section

bhumi1102 committed:
mail object and processing status

bhumi1102 committed:
Merge branch 'rails:main' into bhumi-guides-action-mailbox

bhumi1102 committed:
explain example

bhumi1102 committed:
Merge branch 'bhumi-guides-action-mailbox' of github.com:bhumi1102/rails into bhumi-guides-action-mailbox

bhumi1102 committed:
Merge branch 'rails:main' into bhumi-guides-action-mailbox

bhumi1102 committed:
mail object code snippet

bhumi1102 committed:
title

bhumi1102 committed:
review comments

Co-authored-by: Ridhwana <Ridhwana.Khan16@gmail.com>

bhumi1102 committed:
review comments

Co-authored-by: Ridhwana <Ridhwana.Khan16@gmail.com>

bhumi1102 committed:
review comments

Co-authored-by: Carlos Antonio da Silva <carlosantoniodasilva@gmail.com>

bhumi1102 committed:
review comments

Co-authored-by: Petrik de Heus <petrik@deheus.net>

bhumi1102 committed:
review comments

Co-authored-by: Petrik de Heus <petrik@deheus.net>

bhumi1102 committed:
review comments

bhumi1102 committed:
review comments

Co-authored-by: Carlos Antonio da Silva <carlosantoniodasilva@gmail.com>

bhumi1102 committed:
Apply suggestions from code review

Co-authored-by: Carlos Antonio da Silva <carlosantoniodasilva@gmail.com>

bhumi1102 committed:
linking

bhumi1102 committed:
review comments

bhumi1102 committed:
explain routing

Co-authored-by: Carlos Antonio da Silva <carlosantoniodasilva@gmail.com>

bhumi1102 committed:
links

bhumi1102 committed:
cleanup

bhumi1102 committed:
callback

bhumi1102 committed:
cleanup

bhumi1102 committed:
cleanup

bhumi1102 committed:
linter

bhumi1102 committed:
typo

Co-authored-by: Carlos Antonio da Silva <carlosantoniodasilva@gmail.com>

bhumi1102 committed:
review comments

Co-authored-by: Carlos Antonio da Silva <carlosantoniodasilva@gmail.com>

bhumi1102 committed:
review comments

Co-authored-by: Carlos Antonio da Silva <carlosantoniodasilva@gmail.com>

bhumi1102 committed:
review comments

Co-authored-by: Carlos Antonio da Silva <carlosantoniodasilva@gmail.com>

bhumi1102 committed:
cleanup

bhumi1102 committed:
add routing docs

bhumi1102 committed:
linter

Schwad commented on guides/source/action_mailbox_basics.md:
```diff
@@ -18,30 +18,41 @@ After reading this guide, you will know:
-processing in Rails. It ships with ingresses for Mailgun, Mandrill, Postmark,
-and SendGrid. You can also handle inbound mails directly via the built-in Exim,
-Postfix, and Qmail ingresses.
+Action Mailbox routes incoming emails to controller-like mailboxes for processing in your Rails application. Action Mailbox is for receiving email, while [Action Mailer](action_mailer_basics.html) is for *sending* them.
```
I really like how this streamlines the introduction to Action Mailbox and clarifying the difference to Action Mailer! 

Something I think could have tripped up developers in the past.

bhumi1102 replied:
Yes exactly! Since they both have to do with emails.

Schwad commented on guides/source/action_mailbox_basics.md:
```diff
@@ -18,30 +18,41 @@ After reading this guide, you will know:
 ## Setup
 
-Install migrations needed for `InboundEmail` and ensure Active Storage is set up:
+Action Mailbox has a few moving parts. First you'll run the installer. Then configure a chosen ingress for handling incoming email. You're then ready to add Action Mailbox routing, create mailboxes and start processing incoming emails.
```
Great explainer of what the setup means before walking through.

tjschuck replied:
```suggestion
Action Mailbox has a few moving parts. First, you'll run the installer. Next, you'll choose and configure an ingress for handling incoming email. You're then ready to add Action Mailbox routing, create mailboxes, and start processing incoming emails.
```

Schwad commented on guides/source/action_mailbox_basics.md:
````diff
@@ -277,13 +294,62 @@ class ApplicationMailbox < ActionMailbox::Base
+
+When processing an `InboundEmail`, you can get the parsed version of the email as a [`Mail`](https://github.com/mikel/mail) object with `InboundEmail#mail`. You can also get the raw source directly using the `#source` method. With `Mail` object, you can access the relevant fields, such as `mail.to`, `mail.body.decoded`, etc.
+
+```bash
````
I love seeing this code walkthrough explicitly, very helpful.

bhumi1102 replied:
The README of that `mail` library was good read :)

Schwad commented on guides/source/action_mailbox_basics.md:
```diff
@@ -277,13 +294,62 @@ class ApplicationMailbox < ActionMailbox::Base
+
+Once an email has been routed to a matching mailbox and processed, Action Mailbox updates the email status stored in `action_mailbox_inbound_emails` table with one of the following values:
+
+- `pending`: Just received by one of the ingress controllers and scheduled for routing.
```
Would it read more cleanly with "Just" removed?

Schwad reviewed (commented):
I did not review this from a "correctness" perspective (eg testing and critiquing the code locally) but as a documentation read through as a developer.

I really think these changes all help improve existing documentation on each change and also add additional helpful context I didn't know. I learned a bit!

Really liked the new code block walkthroughs.

bhumi1102 committed:
community review

tjschuck commented on guides/source/action_mailbox_basics.md:
```diff
@@ -277,13 +294,62 @@ class ApplicationMailbox < ActionMailbox::Base
+- `failed`: An exception was raised during the specific mailbox’s execution of the `process` method.
+- `bounced`: Rejected processing by the specific mailbox and bounced to sender.
+
+If the email is marked either `delivered`, `failed`, or `bounced` it's considered 'processed' and marked for [incineration](#incineration-of-inboundemails).
```
```suggestion
If the email is marked either `delivered`, `failed`, or `bounced`, it's considered "processed" and marked for [incineration](#incineration-of-inboundemails).
```

tjschuck commented on guides/source/action_mailbox_basics.md:
```diff
@@ -277,13 +294,62 @@ class ApplicationMailbox < ActionMailbox::Base
+
+## Example
+
+Here is an example of an Action Mailbox that processes emails to create 'forwards' for the user's project.
```
```suggestion
Here is an example of an Action Mailbox that processes emails to create "forwards" for the user's project.
```

tjschuck commented on guides/source/action_mailbox_basics.md:
```diff
@@ -18,30 +18,41 @@ After reading this guide, you will know:
-Install migrations needed for `InboundEmail` and ensure Active Storage is set up:
+Action Mailbox has a few moving parts. First you'll run the installer. Then configure a chosen ingress for handling incoming email. You're then ready to add Action Mailbox routing, create mailboxes and start processing incoming emails.
+
+To start let's install Action Mailbox:
```
```suggestion
To start, let's install Action Mailbox:
```

tjschuck commented on guides/source/action_mailbox_basics.md:
```diff
@@ -18,30 +18,41 @@ After reading this guide, you will know:
-## Configuration
+This will run the Action Mailbox and Active Storage migrations.
+
+The Action Mailbox table `action_mailbox_inbound_emails` stores incoming messages and their processing 'status'.
```
```suggestion
The Action Mailbox table `action_mailbox_inbound_emails` stores incoming messages and their processing status.
```

tjschuck commented on guides/source/action_mailbox_basics.md:
```diff
@@ -18,30 +18,41 @@ After reading this guide, you will know:
+
+The Action Mailbox table `action_mailbox_inbound_emails` stores incoming messages and their processing 'status'.
+
+At this point you can start your rails server and check out `http://localhost:3000/rails/conductor/action_mailbox/inbound_emails`. See [Local Development and Testing](#local-development-and-testing) for more.
```
```suggestion
At this point, you can start your Rails server and check out `http://localhost:3000/rails/conductor/action_mailbox/inbound_emails`. See [Local Development and Testing](#local-development-and-testing) for more.
```

tjschuck commented on guides/source/action_mailbox_basics.md:
```diff
@@ -265,9 +276,15 @@ https://actionmailbox:PASSWORD@example.com/rails/action_mailbox/sendgrid/inbound
+
+### Configure Routing
+
+After an incoming email is received via the configured ingress, it needs to be forwarded to a mailbox for actual processing by your application. Much like the [Rails router](routing.html) that dispatches URLs to controllers, routing in Action Mailbox defines which emails go to which mailboxes for processing. Routes are added to the `application_mailbox.rb` file using regular expression:
```
```suggestion
After an incoming email is received via the configured ingress, it needs to be forwarded to a mailbox for actual processing by your application. Much like the [Rails router](routing.html) that dispatches URLs to controllers, routing in Action Mailbox defines which emails go to which mailboxes for processing. Routes are added to the `application_mailbox.rb` file using regular expressions:
```

tjschuck commented on guides/source/action_mailbox_basics.md:
````diff
@@ -277,13 +294,62 @@ class ApplicationMailbox < ActionMailbox::Base
 ```
 
-Then set up a mailbox:
+The regular expression matches the incoming email's `to`, `cc`, or `bcc` fields. For example, the above will match any email sent to `save@` to a 'forwards' mailbox. There are other ways to route an email, see [`ActionMailbox::Base API`](https://api.rubyonrails.org/classes/ActionMailbox/Base.html) for more.
````
```suggestion
The regular expression matches the incoming email's `to`, `cc`, or `bcc` fields. For example, the above will match any email sent to `save@` to a "forwards" mailbox. There are other ways to route an email; see [`ActionMailbox::Base API`](https://api.rubyonrails.org/classes/ActionMailbox/Base.html) for more.
```

tjschuck commented on guides/source/action_mailbox_basics.md:
```diff
@@ -277,13 +294,62 @@ class ApplicationMailbox < ActionMailbox::Base
-Then set up a mailbox:
+The regular expression matches the incoming email's `to`, `cc`, or `bcc` fields. For example, the above will match any email sent to `save@` to a 'forwards' mailbox. There are other ways to route an email, see [`ActionMailbox::Base API`](https://api.rubyonrails.org/classes/ActionMailbox/Base.html) for more.
+
+We need to create that 'forwards' mailbox next.
```
```suggestion
We need to create that "forwards" mailbox next.
```

tjschuck commented on guides/source/action_mailbox_basics.md:
```diff
@@ -277,13 +294,62 @@ class ApplicationMailbox < ActionMailbox::Base
+
+### Process Email
+
+When processing an `InboundEmail`, you can get the parsed version of the email as a [`Mail`](https://github.com/mikel/mail) object with `InboundEmail#mail`. You can also get the raw source directly using the `#source` method. With `Mail` object, you can access the relevant fields, such as `mail.to`, `mail.body.decoded`, etc.
```
```suggestion
When processing an `InboundEmail`, you can get the parsed version of the email as a [`Mail`](https://github.com/mikel/mail) object with `InboundEmail#mail`. You can also get the raw source directly using the `#source` method. With the `Mail` object, you can access the relevant fields, such as `mail.to`, `mail.body.decoded`, etc.
```

tjschuck commented on guides/source/action_mailbox_basics.md:
```diff
@@ -380,3 +423,13 @@ end
+
+Default data incineration ensures that you're not holding on to people's data unnecessarily after they may have canceled their accounts or deleted their content.
+
+The intention with Action Mailbox processing is that as you process an email, you should extract all the data you need from the email and persist into domain models in your application. The `InboundEmail` simply stays in the system for the extra time to allow for debugging and forensic and then will be deleted.
```
```suggestion
The intention with Action Mailbox processing is that as you process an email, you should extract all the data you need from the email and persist it into domain models in your application. The `InboundEmail` stays in the system for the configured time to allow for debugging and forensics and then will be deleted.
```

tjschuck commented on guides/source/action_mailbox_basics.md:
```diff
@@ -380,3 +423,13 @@ end
+
+## Incineration of InboundEmails
+
+By default, an `InboundEmail` that has been processed will be incinerated after 30 days. The InboundEmail is considered as processed when its status changes to `delivered`, `failed` or `bounced`.
```
```suggestion
By default, an `InboundEmail` that has been processed will be incinerated after 30 days. The InboundEmail is considered as processed when its status changes to `delivered`, `failed`, or `bounced`.
```

tjschuck commented on guides/source/action_mailbox_basics.md:
```diff
@@ -277,13 +294,62 @@ class ApplicationMailbox < ActionMailbox::Base
+
+The `before_processing` callback is used to ensure that certain conditions are met before `process` method is called. In this case, `before_processing` checks that the user has at least one project. Other supported [Action Mailbox callbacks](https://api.rubyonrails.org/classes/ActionMailbox/Callbacks.html) are `after_processing` and `around_processing`.
+
+The email can be bounced using `bounced_with` if the 'forwarder' has no projects. The 'forwarder' is a `User` with the same email as `mail.from`.
```
```suggestion
The email can be bounced using `bounced_with` if the "forwarder" has no projects. The "forwarder" is a `User` with the same email as `mail.from`.
```

tjschuck commented on guides/source/action_mailbox_basics.md:
```diff
@@ -277,13 +294,62 @@ class ApplicationMailbox < ActionMailbox::Base
+
+The email can be bounced using `bounced_with` if the 'forwarder' has no projects. The 'forwarder' is a `User` with the same email as `mail.from`.
+
+If the 'forwarder' does have at least one project, the `record_forward` method creates an Active Record model in the application using the email data `mail.subject` and `mail.content`. Otherwise it sends an email, using Action Mailer, requesting the 'forwarder' to choose a project.
```
```suggestion
If the "forwarder" does have at least one project, the `record_forward` method creates an Active Record model in the application using the email's `mail.subject` and `mail.content`. Otherwise, it sends an email using Action Mailer that requests that the "forwarder" choose a project.
```

lazyatom commented:
I realise this isn't a part of the guide that you've changed, but reading through it, I noticed in the main code example that the method `Mail#content` is called (within the `record_forward` method in the example). However, as far as I can tell, this method doesn't actually exist. If I'm wrong, apologies, but if I'm right, might be worth addressing that here too?

bhumi1102 committed:
community review comments

Co-authored-by: T.J. Schuck <tj@tjschuck.com>

bhumi1102 committed:
community review comments

Co-authored-by: T.J. Schuck <tj@tjschuck.com>

bhumi1102 committed:
community review comments

Co-authored-by: T.J. Schuck <tj@tjschuck.com>

bhumi1102 committed:
community review comments

Co-authored-by: T.J. Schuck <tj@tjschuck.com>

bhumi1102 committed:
community review comments

Co-authored-by: T.J. Schuck <tj@tjschuck.com>

bhumi1102 committed:
community review comments

Co-authored-by: T.J. Schuck <tj@tjschuck.com>

bhumi1102 commented:
> I realise this isn't a part of the guide that you've changed, but reading through it, I noticed in the main code example that the method `Mail#content` is called (within the `record_forward` method in the example). However, as far as I can tell, this method doesn't actually exist. If I'm wrong, apologies, but if I'm right, might be worth addressing that here too?

@lazyatom You're right.

```
`(byebug) mail.body.decoded
"Does it have a content field? and is it mail.decoded or mail.body.decoded?"
(byebug) mail.decoded
"Does it have a content field? and is it mail.decoded or mail.body.decoded?"
(byebug) mail.content
*** NoMethodError Exception: undefined method `content' for #<Mail::Message:61780, Multipart: false, Headers: <Date: Mon, 12 Feb 2024 11:05:1`
```

I value this type of feedback. Thanks for catching, will fix.

bhumi1102 committed:
fix code examples issue with mail.content

bhumi1102 committed:
community review comments

Co-authored-by: T.J. Schuck <tj@tjschuck.com>

bhumi1102 committed:
community review comments

Co-authored-by: T.J. Schuck <tj@tjschuck.com>

bhumi1102 committed:
type

bhumi1102 committed:
quotes

carlosantoniodasilva commented on guides/source/action_mailbox_basics.md:
```diff
@@ -309,7 +378,7 @@ class ForwardsMailbox < ApplicationMailbox
 
     def record_forward
-      forwarder.forwards.create subject: mail.subject, content: mail.content
+      forwarder.forwards.create subject: mail.subject, content: mail.body.decoded
```
Minor: maybe we just use the shortcut version, `mail.decoded`?

```suggestion
      forwarder.forwards.create subject: mail.subject, content: mail.decoded
```

carlosantoniodasilva commented on guides/source/action_mailbox_basics.md:
```diff
@@ -277,13 +294,65 @@ class ApplicationMailbox < ActionMailbox::Base
+
+The email can be bounced using `bounced_with` if the "forwarder" has no projects. The "forwarder" is a `User` with the same email as `mail.from`.
+
+If the "forwarder" does have at least one project, the `record_forward` method creates an Active Record model in the application using the email data `mail.subject` and `mail.body.decoded`. Otherwise, it sends an email, using Action Mailer, requesting the "forwarder" to choose a project.
```
Minor: maybe we just use the shortcut version, `mail.decoded`?
```suggestion
If the "forwarder" does have at least one project, the `record_forward` method creates an Active Record model in the application using the email data `mail.subject` and `mail.decoded`. Otherwise, it sends an email, using Action Mailer, requesting the "forwarder" to choose a project.
```

carlosantoniodasilva commented on guides/source/action_mailbox_basics.md:
```diff
@@ -374,9 +420,19 @@ class ForwardsMailboxTest < ActionMailbox::TestCase
     assert_equal people(:david), recording.creator
     assert_equal "Status update?", recording.forward.subject
-    assert_match "What's the status?", recording.forward.content.to_s
+    assert_match "What's the status?", recording.forward.decoded.to_s
```
I believe this should still be `content`, as it's related to the `forward` Active Record model, `content` column. (Not the mail object) That's how the example above stores it: `content: mail.body.decoded` (or `content: mail.decoded` if we use the shortcut method mentioned as well)

```suggestion
    assert_match "What's the status?", recording.forward.content.to_s
```

carlosantoniodasilva commented on guides/source/action_mailbox_basics.md:
```diff
@@ -374,9 +420,19 @@ class ForwardsMailboxTest < ActionMailbox::TestCase
+
+## Incineration of InboundEmails
+
+By default, an `InboundEmail` that has been processed will be incinerated after 30 days. The InboundEmail is considered as processed when its status changes to `delivered`, `failed`, or `bounced`.
```
This is the only instance where `InboudEmail` is not shown as code, let's wrap it too for consistency with the rest, since it's actually referring to the object in code:

```suggestion
By default, an `InboundEmail` that has been processed will be incinerated after 30 days. The `InboundEmail` is considered as processed when its status changes to `delivered`, `failed`, or `bounced`.
```

Alternatively, we can consider skipping the second mention in the paragraph:

```suggestion
By default, an `InboundEmail` that has been processed will be incinerated after 30 days. It is considered as processed when its status changes to `delivered`, `failed`, or `bounced`.
```

carlosantoniodasilva reviewed (commented):
Really great updates, just had a couple other minor suggestions and one potential code fix, otherwise I think it's good to go.

bhumi1102 committed:
last review

Co-authored-by: Carlos Antonio da Silva <carlosantoniodasilva@gmail.com>

bhumi1102 committed:
last review

Co-authored-by: Carlos Antonio da Silva <carlosantoniodasilva@gmail.com>

bhumi1102 committed:
last review

Co-authored-by: Carlos Antonio da Silva <carlosantoniodasilva@gmail.com>

bhumi1102 committed:
last review

Co-authored-by: Carlos Antonio da Silva <carlosantoniodasilva@gmail.com>

bhumi1102 commented:
> Really great updates, just had a couple other minor suggestions and one potential code fix, otherwise I think it's good to go.

Done!

tjschuck commented on guides/source/action_mailbox_basics.md:
````diff
@@ -277,13 +294,65 @@ class ApplicationMailbox < ActionMailbox::Base
+
+When processing an `InboundEmail`, you can get the parsed version of the email as a [`Mail`](https://github.com/mikel/mail) object with `InboundEmail#mail`. You can also get the raw source directly using the `#source` method. With the `Mail` object, you can access the relevant fields, such as `mail.to`, `mail.body.decoded`, etc.
+
+```bash
+$ mail
+#<Mail::Message:33780, Multipart: false, Headers: <Date: Wed, 31 Jan 2024 22:18:40 -0600>, <From: someone@hey.com>, <To: save@example.com>, <Message-ID: <65bb1ba066830_50303a70397e@Bhumis-MacBook-Pro.local.mail>>, <In-Reply-To: >, <Subject: Hello Action Mailbox>, <Mime-Version: 1.0>, <Content-Type: text/plain; charset=UTF-8>, <Content-Transfer-Encoding: 7bit>, <x-original-to: >>
+$ mail.to
+["save@example.com"]
+$ mail.from
+["someone@hey.com"]
+$ mail.date
+Wed, 31 Jan 2024 22:18:40 -0600
+$ mail.subject
+"Hello Action Mailbox"
+$ mail.body
+#<Mail::Body:0x00007fc74cbf46c0 @boundary=nil, @preamble=nil, @epilogue=nil, @charset="US-ASCII", @part_sort_order=["text/plain", "text/enriched", "text/html", "multipart/alternative"], @parts=[], @raw_source="This is the body of the email message.", @ascii_only=true, @encoding="7bit">
+$ mail.body.decoded
+"This is the body of the email message."
+# mail.decoded, a shorthand for mail.body.decoded, also works
+$ mail.decoded
+"This is the body of the email message."
+```
````
This is IRB, not bash.  Other guides [use an IRB code fence](https://github.com/rails/rails/blob/6e7ef7d61c7146ca03b173abc32f7ed97e3d949a/guides/source/active_record_querying.md?plain=1#L172-L176).

```suggestion
```irb
irb> mail
=> #<Mail::Message:33780, Multipart: false, Headers: <Date: Wed, 31 Jan 2024 22:18:40 -0600>, <From: someone@hey.com>, <To: save@example.com>, <Message-ID: <65bb1ba066830_50303a70397e@Bhumis-MacBook-Pro.local.mail>>, <In-Reply-To: >, <Subject: Hello Action Mailbox>, <Mime-Version: 1.0>, <Content-Type: text/plain; charset=UTF-8>, <Content-Transfer-Encoding: 7bit>, <x-original-to: >>
irb> mail.to
=> ["save@example.com"]
irb> mail.from
=> ["someone@hey.com"]
irb> mail.date
=> Wed, 31 Jan 2024 22:18:40 -0600
irb> mail.subject
=> "Hello Action Mailbox"
irb> mail.body
=> #<Mail::Body:0x00007fc74cbf46c0 @boundary=nil, @preamble=nil, @epilogue=nil, @charset="US-ASCII", @part_sort_order=["text/plain", "text/enriched", "text/html", "multipart/alternative"], @parts=[], @raw_source="This is the body of the email message.", @ascii_only=true, @encoding="7bit">
irb> mail.body.decoded
=> "This is the body of the email message."
# mail.decoded, a shorthand for mail.body.decoded, also works
irb> mail.decoded
=> "This is the body of the email message."
```
```

bhumi1102 committed:
last review

Co-authored-by: T.J. Schuck <tj@tjschuck.com>

bhumi1102 committed:
wrap column at 80

carlosantoniodasilva committed:
Clarify that email status is updated as the email gets routed/processed

The status changes as the email goes through the various steps.

carlosantoniodasilva committed:
Remove `API` from the link text to fix bad wrapping

When generating the content, it was forcing a line break between
ActionMailbox::Base and API, causing the output HTML content to be
weird. Removing API and wrapping all content again at 80 columns fixes
it.

carlosantoniodasilva committed:
Revert first line break to keep it consistent
